File: examples/Delphi/Ex2DArrU.pas
(* ***** BEGIN LICENSE BLOCK *****
 * Version: MPL 1.1
 *
 * The contents of this file are subject to the Mozilla Public License Version
 * 1.1 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * The Original Code is TurboPower SysTools
 *
 * The Initial Developer of the Original Code is
 * TurboPower Software
 *
 * Portions created by the Initial Developer are Copyright (C) 1996-2002
 * the Initial Developer. All Rights Reserved.
 *
 * Contributor(s):
 *
 * ***** END LICENSE BLOCK ***** *)

unit Ex2DArrU;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls,
  Forms, Dialogs, StdCtrls, Buttons, ExtCtrls,

  StConst, StUtils, StBase, StLArr;

type
  TSTDlg = class(TForm)
    ArrayLB: TListBox;
    CreateBtn: TButton;
    Label5: TLabel;
    LMValue: TEdit;
    Label6: TLabel;
    LMRow: TEdit;
    LMCol: TEdit;
    ClearBtn: TButton;
    FillBtn: TButton;
    PutBtn: TButton;
    PutRowBtn: TButton;
    GetBtn: TButton;
    GetRowBtn: TButton;
    SortBtn: TButton;
    LoadBtn: TButton;
    SaveBtn: TButton;
    OD1: TOpenDialog;
    SD1: TSaveDialog;

    procedure FormCreate(Sender: TObject);
    procedure FormClose(Sender: TObject; var Action: TCloseAction);

    procedure CreateBtnClick(Sender: TObject);
    procedure ClearBtnClick(Sender: TObject);
    procedure FillBtnClick(Sender: TObject);
    procedure PutBtnClick(Sender: TObject);
    procedure GetBtnClick(Sender: TObject);
    procedure PutRowBtnClick(Sender: TObject);
    procedure GetRowBtnClick(Sender: TObject);
    procedure SortBtnClick(Sender: TObject);
    procedure LoadBtnClick(Sender: TObject);
    procedure SaveBtnClick(Sender: TObject);
    procedure ArrayLBDblClick(Sender: TObject);
  private
    { Private declarations }
  public
    { Public declarations }
    procedure SetBusy(B : Boolean);
    procedure FillListBox;
    procedure UpdateButtons(AOK : Boolean);
  end;

var
  STDlg: TSTDlg;

implementation

{$R *.DFM}

type
  S10 = string[10];

const
  MaxRows = 1000;
  MaxCols = 10;

var
  MyLMatrix : TStLMatrix;
  LIArray : array[1..MaxCols] of LongInt;

function MyArraySort(const E1, E2) : Integer; far;
begin
  Result := LongInt(E1) - LongInt(E2);
end;

procedure TSTDlg.UpdateButtons(AOK : Boolean);
begin
  ClearBtn.Enabled  := AOK;
  FillBtn.Enabled   := AOK;
  PutBtn.Enabled    := AOK;
  PutRowBtn.Enabled := AOK;
  GetBtn.Enabled    := AOK;
  GetRowBtn.Enabled := AOK;
  SortBtn.Enabled   := AOK;
  SaveBtn.Enabled   := AOK;
end;

procedure TSTDlg.SetBusy(B : Boolean);
begin
  if B then
    Screen.Cursor := crHourGlass
  else
    Screen.Cursor := crDefault;
end;

procedure TSTDlg.FormCreate(Sender: TObject);
begin
  RegisterClass(TStLMatrix);
  UpdateButtons(False);
end;

procedure TSTDlg.FormClose(Sender: TObject;
  var Action: TCloseAction);
begin
  MyLMatrix.Free;
end;

procedure TSTDlg.FillListBox;
var
  row,
  col,
  Value : LongInt;
begin
  SetBusy(True);
  ArrayLB.Clear;
  ArrayLB.Perform(WM_SETREDRAW,0,0);

  for row := 0 to MyLMatrix.Rows-1 do
  begin
    for col := 0 to MyLMatrix.Cols-1 do
    begin
      MyLMatrix.Get(row,col,Value);
      ArrayLB.Items.Add(IntToStr(row) + ',' +
                        IntToStr(col) + ' = ' + IntToStr(Value));
    end;
  end;
  ArrayLB.Perform(WM_SETREDRAW,1,0);
  ArrayLB.Update;
  SetBusy(False);
end;


procedure TSTDlg.CreateBtnClick(Sender: TObject);
var
  row,
  col,
  Value  : LongInt;
begin
  ArrayLB.Clear;

  if Assigned(MyLMatrix) then
    MyLMatrix.Free;

  UpdateButtons(False);
  MyLMatrix := TStLMatrix.Create(MaxRows,MaxCols,sizeof(LongInt));
  MyLMatrix.ElementsStorable := True;

  SetBusy(True);
  for row := 0 to MaxRows-1 do
  begin
    for col := 0 to MaxCols-1 do
    begin
      Value := Trunc(Random(10000));
      MyLMatrix.Put(row,col,Value);
    end;
  end;
  SetBusy(False);

  FillListBox;
  UpdateButtons(True);

  LMRow.Text := '0';
  LMCol.Text := '0';
  MyLMatrix.Get(0,0,Value);
  LMValue.Text := IntToStr(Value);
end;

procedure TSTDlg.ClearBtnClick(Sender: TObject);
var
  Value : LongInt;
begin
  MyLMatrix.Clear;
  ArrayLB.Clear;

  LMRow.Text := '0';
  LMCol.Text := '0';
  MyLMatrix.Get(0,0,Value);
  LMValue.Text := IntToStr(Value);
end;

procedure TSTDlg.FillBtnClick(Sender: TObject);
var
  row,
  col,
  Value : LongInt;
begin
  if (LMValue.Text = '') then
  begin
    ShowMessage('No value entered');
    Exit;
  end;

  Value := StrToInt(LMValue.Text);
  MyLMatrix.Fill(Value);

  FillListBox;

  row := 0;
  col := 0;
  LMRow.Text := IntToStr(row);
  LMCol.Text := IntToStr(col);

  MyLMatrix.Get(row, col, Value);
  LMValue.Text := IntToStr(Value);

  SetBusy(False);
end;

procedure TSTDlg.PutBtnClick(Sender: TObject);
var
  LBE,
  row,
  col,
  Value : LongInt;
begin
  if (LMValue.Text = '') then
  begin
    ShowMessage('No value entered');
    Exit;
  end;

  if (LMRow.Text = '') then
    LMRow.Text := '0';
  if (LMCol.Text = '') then
    LMCol.Text := '0';

  Value := StrToInt(LMValue.Text);
  row := StrToInt(LMRow.Text);
  col := StrToInt(LMCol.Text);
  MyLMatrix.Put(row,col,Value);

  LBE := (row * MaxRows) + col;
  ArrayLB.Items[LBE] := IntToStr(row) + ',' +
                        IntToStr(col) + ' = ' + IntToStr(Value);

  row := StrToInt(LMRow.Text);
  col := StrToInt(LMCol.Text);
  MyLMatrix.Get(row, col, Value);
  LMValue.Text := IntToStr(Value);
end;

procedure TSTDlg.GetBtnClick(Sender: TObject);
var
  LBE,
  row,
  col,
  Value : LongInt;
begin
  if (LMValue.Text = '') then begin
    ShowMessage('No value entered');
    Exit;
  end;

  if (LMRow.Text = '') then
    LMRow.Text := '0';
  if (LMCol.Text = '') then
    LMCol.Text := '0';

  Value := StrToInt(LMValue.Text);
  row := StrToInt(LMRow.Text);
  col := StrToInt(LMCol.Text);
  MyLMatrix.Get(row,col,Value);

  LMRow.Text := IntToStr(row);
  LMCol.Text := IntToStr(col);
  LMValue.Text := IntToStr(Value);

  LBE := (row * MaxCols) + col;
  ArrayLB.ItemIndex := LBE;
end;

procedure TSTDlg.PutRowBtnClick(Sender: TObject);
var
  row,
  col,
  Value : LongInt;

begin
  if (LMValue.Text = '') then
  begin
    ShowMessage('No value entered');
    Exit;
  end;

  if (LMRow.Text = '') then
    LMRow.Text := '0';

  Value := StrToInt(LMValue.Text);
  row := StrToInt(LMRow.Text);

  FillStruct(LIArray,MaxCols,Value,SizeOf(Value));

  MyLMatrix.PutRow(row,LIArray);
  FillListBox;

  row := StrToInt(LMRow.Text);
  col := 0;
  MyLMatrix.Get(row, col, Value);

  LMValue.Text := IntToStr(Value);
  LMCol.Text := '0';
end;

procedure TSTDlg.GetRowBtnClick(Sender: TObject);
var
  step,
  LIV   : LongInt;

begin
  if (LMRow.Text = '') then
    LMRow.Text := '0';

  LIV := 0;
  FillStruct(LIArray,MaxCols,LIV,SizeOf(LIV));
  MyLMatrix.GetRow(0,LIArray);

  ArrayLB.Clear;
  ArrayLB.Perform(WM_SETREDRAW,0,0);

  for step := 1 to MaxCols do
    ArrayLB.Items.Add('Col' + IntToStr(step-1) + ': ' + IntToStr(LIArray[step]));

  ArrayLB.Perform(WM_SETREDRAW,1,0);
  ArrayLB.Update;
end;

procedure TSTDlg.SortBtnClick(Sender: TObject);
begin
  MyLMatrix.SortRows(0,MyArraySort);
  FillListBox;
end;

procedure TSTDlg.LoadBtnClick(Sender: TObject);
begin
  if (OD1.Execute) then
  begin
    if NOT (Assigned(MyLMatrix)) then
    begin
      UpdateButtons(False);
      MyLMatrix := TStLMatrix.Create(MaxRows,MaxCols,sizeof(LongInt));
      MyLMatrix.ElementsStorable := True;
    end;
    MyLMatrix.LoadFromFile(OD1.FileName);
    FillListBox;
    UpdateButtons(True);
  end;
end;

procedure TSTDlg.SaveBtnClick(Sender: TObject);
begin
  if (SD1.Execute) then
    MyLMatrix.StoreToFile(SD1.FileName);
end;

procedure TSTDlg.ArrayLBDblClick(Sender: TObject);
var
  row,
  col,
  I,
  Value  : LongInt;

begin
  I := ArrayLB.ItemIndex;
  row := I div MaxCols;
  col := I mod MaxCols;

  MyLMatrix.Get(row, col, Value);
  LMRow.Text := IntToStr(row);
  LMCol.Text := IntToStr(col);
  LMValue.Text := IntToStr(Value);
end;



end.

File: examples/CBuilder/Ex2DArrU.cpp
// ***** BEGIN LICENSE BLOCK *****
// * Version: MPL 1.1
// *
// * The contents of this file are subject to the Mozilla Public License Version
// * 1.1 (the "License"); you may not use this file except in compliance with
// * the License. You may obtain a copy of the License at
// * http://www.mozilla.org/MPL/
// *
// * Software distributed under the License is distributed on an "AS IS" basis,
// * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
// * for the specific language governing rights and limitations under the
// * License.
// *
// * The Original Code is TurboPower SysTools
// *
// * The Initial Developer of the Original Code is
// * TurboPower Software
// *
// * Portions created by the Initial Developer are Copyright (C) 1996-2002
// * the Initial Developer. All Rights Reserved.
// *
// * Contributor(s):
// *
// * ***** END LICENSE BLOCK *****
//---------------------------------------------------------------------------
#include <vcl\vcl.h>
#pragma hdrstop

#include <stdlib.h>
#include "Ex2DArrU.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "StLArr"
#pragma link "StUtils"

#pragma resource "*.dfm"


TForm1 *Form1;
//---------------------------------------------------------------------------
__fastcall TForm1::TForm1(TComponent* Owner)
	: TForm(Owner)
{
}
//---------------------------------------------------------------------------
int __fastcall MyArraySort(const void* E1, const void* E2)
{
  return *(long*)E1 - *(long*)E2;
}
//---------------------------------------------------------------------------

void __fastcall TForm1::FormCreate(TObject *Sender)
{
  //RegisterClass should be used, but it is not there in C++Builder 1.0
  RegisterClassAlias(__classid(TStLMatrix), "TStLMatrix");
  UpdateButtons(false);
}
//---------------------------------------------------------------------------
void __fastcall TForm1::FormClose(TObject *Sender, TCloseAction &Action)
{
  delete MyLMatrix;
}
//---------------------------------------------------------------------------
void TForm1::UpdateButtons(bool AOK)
{
  ClearBtn->Enabled  = AOK;
  FillBtn->Enabled   = AOK;
  PutBtn->Enabled    = AOK;
  PutRowBtn->Enabled = AOK;
  GetBtn->Enabled    = AOK;
  GetRowBtn->Enabled = AOK;
  SortBtn->Enabled   = AOK;
  SaveBtn->Enabled   = AOK;
}
//---------------------------------------------------------------------------
void TForm1::SetBusy(bool B)
{
  if (B)
    Screen->Cursor = crHourGlass;
  else
    Screen->Cursor = crDefault;
}
//---------------------------------------------------------------------------
void TForm1::FillListBox()
{
  long Value;

  SetBusy(true);
  ArrayLB->Clear();
  ArrayLB->Perform(WM_SETREDRAW, 0, 0);

  for (Cardinal Row = 0; Row < MyLMatrix->Rows; Row++) {
    for (Cardinal Col = 0; Col < MyLMatrix->Cols; Col++) {
      MyLMatrix->Get(Row, Col, &Value);
      ArrayLB->Items->Add(IntToStr(Row) + "," +
                        IntToStr(Col) + " = " + IntToStr((int)Value));
    };
  };

  ArrayLB->Perform(WM_SETREDRAW,1,0);
  ArrayLB->Update();
  SetBusy(false);
}
//---------------------------------------------------------------------------
void __fastcall TForm1::CreateBtnClick(TObject *Sender)
{
  long Value;

  ArrayLB->Clear();

  if (MyLMatrix)
    delete MyLMatrix;

  UpdateButtons(false);
  MyLMatrix = new TStLMatrix(MAXROWS, MAXCOLS, sizeof(long));
  MyLMatrix->ElementsStorable = true;

  SetBusy(true);
  for (Cardinal Row = 0; Row < MAXROWS; Row++) {
    for (Cardinal Col = 0; Col < MAXCOLS; Col++) {
      Value = floor(random(10000));
      MyLMatrix->Put(Row, Col, &Value);
    };
  };
  SetBusy(false);

  FillListBox();
  UpdateButtons(true);

  LMRow->Text = "0";
  LMCol->Text = "0";
  MyLMatrix->Get(0, 0, &Value);
  LMValue->Text = IntToStr((int)Value);
}
//---------------------------------------------------------------------------
void __fastcall TForm1::ClearBtnClick(TObject *Sender)
{
  long Value;

  MyLMatrix->Clear();
  ArrayLB->Clear();

  LMRow->Text = "0";
  LMCol->Text = "0";
  MyLMatrix->Get(0, 0, &Value);
  LMValue->Text = IntToStr((int)Value);
}
//---------------------------------------------------------------------------
void __fastcall TForm1::FillBtnClick(TObject *Sender)
{
  long Row, Col, Value;

  if (LMValue->Text == "") {
    ShowMessage("No value entered");
    return;
  };

  Value = StrToInt(LMValue->Text);
  MyLMatrix->Fill(&Value);

  FillListBox();

  Row = 0;
  Col = 0;
  LMRow->Text = IntToStr((int)Row);
  LMCol->Text = IntToStr((int)Col);

  MyLMatrix->Get(Row, Col, &Value);
  LMValue->Text = IntToStr((int)Value);

  SetBusy(false);
}
//---------------------------------------------------------------------------
void __fastcall TForm1::PutBtnClick(TObject *Sender)
{
  long LBE, Row, Col, Value;

  if (LMValue->Text == "") {
    ShowMessage("No value entered");
    return;
  };

  if (LMRow->Text == "")
    LMRow->Text = "0";
  if (LMCol->Text == "")
    LMCol->Text = "0";

  Value = StrToInt(LMValue->Text);
  Row = StrToInt(LMRow->Text);
  Col = StrToInt(LMCol->Text);
  MyLMatrix->Put(Row, Col, &Value);

  LBE = (Row * MAXROWS) + Col;
  ArrayLB->Items->Strings[LBE] = IntToStr((int)Row) + "," +
                        IntToStr((int)Col) + " = " + IntToStr((int)Value);

  Row = StrToInt(LMRow->Text);
  Col = StrToInt(LMCol->Text);
  MyLMatrix->Get(Row, Col, &Value);
  LMValue->Text = IntToStr((int)Value);
}
//---------------------------------------------------------------------------
void __fastcall TForm1::PutRowBtnClick(TObject *Sender)
{
  long Row, Col, Value;

  if (LMValue->Text == "") {
    ShowMessage("No value entered");
    return;
  };

  if (LMRow->Text == "")
    LMRow->Text = "0";

  Value = StrToInt(LMValue->Text);
  Row = StrToInt(LMRow->Text);

  FillStruct(&LIArray, MAXCOLS, &Value, sizeof(Value));

  MyLMatrix->PutRow(Row, &LIArray);
  FillListBox();

  Row = StrToInt(LMRow->Text);
  Col = 0;
  MyLMatrix->Get(Row, Col, &Value);

  LMValue->Text = IntToStr((int)Value);
  LMCol->Text = "0";
}
//---------------------------------------------------------------------------
void __fastcall TForm1::GetBtnClick(TObject *Sender)
{
  long LBE, Row, Col, Value;

  if (LMValue->Text == "") {
    ShowMessage("No value entered");
    return;
  };

  if (LMRow->Text == "")
    LMRow->Text = "0";
  if (LMCol->Text == "")
    LMCol->Text = "0";

  Value = StrToInt(LMValue->Text);
  Row = StrToInt(LMRow->Text);
  Col = StrToInt(LMCol->Text);
  MyLMatrix->Get(Row, Col, &Value);

  LMRow->Text = IntToStr((int)Row);
  LMCol->Text = IntToStr((int)Col);
  LMValue->Text = IntToStr((int)Value);

  LBE = (Row * MAXCOLS) + Col;
  ArrayLB->ItemIndex = LBE;
}
//---------------------------------------------------------------------------
void __fastcall TForm1::GetRowBtnClick(TObject *Sender)
{
  long LIV;

  if (LMRow->Text == "")
    LMRow->Text = "0";

  LIV = 0;
  FillStruct(&LIArray, MAXCOLS, &LIV, sizeof(LIV));
  MyLMatrix->GetRow(0, &LIArray);

  ArrayLB->Clear();
  ArrayLB->Perform(WM_SETREDRAW,0,0);

  for (long Step = 0; Step < MAXCOLS; Step++)
    ArrayLB->Items->Add("Col" + IntToStr((int)Step) + ": " + IntToStr((int)LIArray[Step]));

  ArrayLB->Perform(WM_SETREDRAW, 1, 0);
  ArrayLB->Update();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::SortBtnClick(TObject *Sender)
{
  MyLMatrix->SortRows(0, MyArraySort);
  FillListBox();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::LoadBtnClick(TObject *Sender)
{
  if (OD1->Execute()) {
    if (!MyLMatrix) {
      UpdateButtons(false);
      MyLMatrix = new TStLMatrix(MAXROWS, MAXCOLS, sizeof(long));
      MyLMatrix->ElementsStorable = true;
    };
    MyLMatrix->LoadFromFile(OD1->FileName);
    FillListBox();
    UpdateButtons(true);
  };
}
//---------------------------------------------------------------------------
void __fastcall TForm1::SaveBtnClick(TObject *Sender)
{
  if (SD1->Execute())
    MyLMatrix->StoreToFile(SD1->FileName);
}
//---------------------------------------------------------------------------
void __fastcall TForm1::ArrayLBDblClick(TObject *Sender)
{
  long Row, Col, I, Value;

  I = ArrayLB->ItemIndex;
  Row = I / MAXCOLS;
  Col = I % MAXCOLS;

  MyLMatrix->Get(Row, Col, &Value);
  LMRow->Text = IntToStr((int)Row);
  LMCol->Text = IntToStr((int)Col);
  LMValue->Text = IntToStr((int)Value);
}
//---------------------------------------------------------------------------
